distribution/lib/Standard/Base/0.0.0-dev/src/Any.enso

```diff
@@ -100,6 +100,9 @@ type Any
                Therefore, there are no more cases to handle in self method.
             _ -> False
 
+    hash_code : Integer
+    hash_code self = @Builtin_Method "Any.hash_code"
```

Comment:
I'd like to point out that the [hashing pivotal](https://www.pivotaltracker.com/n/projects/2539304/stories/181027272) doesn't expect `hash_code` to be exposed to end users. The [custom eq & hash pivotal](https://www.pivotaltracker.com/story/show/1839453280) will provide users a way to override `==` and `hash`, but it is not going to done via `Any.hash_code` method - but rather a _comparator concept_.

As such I'd suggest to remove this `@Builtin_Method "Any.hash_code"`.

Reply:
Right. I guess I got a bit confused and treated it like `Any.==` from [Convert Any.= to a builtin PR](https://github.com/enso-org/enso/pull/3956), where `Any.==` is still needed for some table specific overrides. Before this PR is ready for the final review, I will remember to remove `Any.hash_code`.

Reply:
How will then the user be able to see the hashcode of an object? Or do we want to allow the user to override the hashcode logic but not allow to _use_ it apart from our primitives?

I imagine as a user I want to be able to know a hashcode so that I can write my own hashed data structures if needed. Will I be able to access `obj.hash_code` in some way?

Reply:
> Will I be able to access obj.hash_code in some way?

`obj.comparator.hash_code` should be a valid way. More specifically:
```
if obj.comparator.is_nothing.not then obj.comparator.hash_code else Nothing
```

Reply:
But until we address the hashing pivotal, I will keep `Any.hash_code` for testing purposes, and mark it as PRIVATE and UNSTABLE for clarity. So that I can at least test `hash_code` on this PR

Reply:
Ok, sounds good!

Reply:
> I imagine as a user I want to be able to know a hashcode so that I can write my own hashed data structures if needed. Will I be able to access `obj.hash_code` in some way?

There are no plans to expose `hash_code` to end users. Only `Map` builtin can access the `hash_code` value. To quote the [pivotal](https://www.pivotaltracker.com/story/show/181027272):

> How to hide these gory details? 
> We want all the Enso objects to be able to participate in the Map. We don't want to expose hash code to the developers (unlike Java) and still have the Map (like JavaScript).

I don't want Enso to be [hash oriented language](http://wiki.apidesign.org/wiki/Smalltalk).

> will keep Any.hash_code for testing purposes

Rather not. Delete it. Test the `Map` builtin.

Reply:
> There are no plans to expose hash_code to end users. Only Map builtin can access the hash_code value. To quote the [pivotal](https://www.pivotaltracker.com/story/show/181027272):

I am 100% for not exposing `hash_code` and hiding it behind `Map`, after all, I was the first one to propose that. However, as @radeusgd pointed out, there is one advanced use case that might affect library devs: Having an atom with many fields, and I want the equivalence/`hash_code` to be based only on one of these fields. How should we compute hash code for this field in such a case?

Don't we still want to have `Any.hash_code` / `Meta.hash_code` as private/unstable?

> Rather not. Delete it. Test the Map builtin.
OK

Reply:
How can I override the hash code logic if I do not have access to hash codes of my parts? Do I have to reimplement all of that whenever I try to override equality?

I think I just _need_ hashcode access to be able to override a hashcode for a custom equality operator.

The other thing is - what if users want to write additional advanced data structures for Enso relying on hashing - like a differently approached hashmap or something else that needs it. We just don't allow it? Or require them to reimplement this logic because we do not want to expose it?

Reply:
In particular, our Table library will need to be able to access the hashcode of Enso objects to implement Distinct/GroupBy efficiently. The hashcode is not needed only by the builtin Map, but there are many other data structures that will need it.

Reply:
@JaroslavTulach : I still have some tests that explicitly invoke `Any.hash_code` (in Java, using `ValuesGenerator`), they are very useful, in my opinion, because I am able to model the formula `for all objects x,y : x != y => hash(x) != hash(y) && x == y => hash(x) == hash(y)` in Java in a pretty readable way with a lot of different values. I know that we ultimately want to use `Any.comparator` and that, ideally, we don't want `Any.comparator` to be publicly available, but can we use that at least internally (unstable / private)?

Moreover, as @radeusgd mentioned before, there are some valid uses cases, where some (advanced) Enso users might want to explicitly get a hash code for some atom.

Reply:
No, I am not going to approve introduction of `Any.hash_code`.

> need to be able to access the hashcode of Enso objects to implement Distinct/GroupBy efficiently

No, you don't need `hash_code`. Use the `Map`  builtin to implement `distinct` or `group_by` - that's the whole idea.

> have some tests that explicitly invoke `Any.hash_code`

If you really need tests that get the **internal** hash_code, then make them part of engine tests.

Reply:
1. How can we use the Map builtin from the Java code within the Table library?

2. If I'm creating an object for which I define a custom comparator, let's say I'm defining an unordered Pair - I want `UPair x y == UPair y x`. How do I implement my custom hashcode if I cannot get a hashcode of `x` and `y` within my implementation?

Reply:
After conversation with @JaroslavTulach, we came to a conclusion, that the use case mentioned by @radeusgd is a valid one, and therefore, there will be explicit (private / unstable / expert) access for a hash code of an atom in the future. Most likely by something like `My_Type.comparator.hash_code my_obj`, but this API's design is still not finished.

Reply:
> How can we use the Map builtin from the Java code within the Table library?

Following shall work:
```java
public static void withMap(java.util.Map<Object, Object> map) { ... }
```
and then from Enso `withMap (Map.empty . insert "A" 10 . insert "B" 20)`. At least I hope it works, does it @Akirathan ?

Reply:
> How do I implement my custom hashcode if I cannot get a hashcode of `x` and `y` within my implementation?

There are few approaches to address this requirement:
- expose `Comparator.hash` for all objects as part of [pivotal 183945328](https://www.pivotaltracker.com/n/projects/2539304/stories/183945328) - then one would obtain the right `Comparator` for an object and called `hash`.
- allow certain fields in `Atom` to be excluded from default implementation of `==` and `hash`  - one would then _annotate(?)_ the fields somehow: `UPair @Exclude_From_Equality x y`
- with [pivotal 183945328](https://www.pivotaltracker.com/n/projects/2539304/stories/183945328) one can actually do it with an _always equals_ wrapper - see below.

### Always Equals Wrapper

In fact we don't need to expose `hash` - we need to _suppress it_! Define:

```haskell
type Exclude_From_Equality_And_Hash
    Value x: Any

Exclude_From_Equality_And_Hash.comparator = Always_Equals

type Always_Equals
     compare self x y = Ordering.Equal
     hash self x = 0
```

and then just construct `UPair (Exclude_From_Equality_And_Hash x) y`.

Reply:
> Following shall work:
> public static void withMap(java.util.Map<Object, Object> map) { ... }
> and then from Enso withMap (Map.empty . insert "A" 10 . insert "B" 20). At least I hope it works, does it @Akirathan ?

Yes, that works. Although I have not found a way how to provide an automated test for that. In Java, `Map` is treated as `com.oracle.truffle.polyglot.PolyglotMap` which extends `java.util.AbstractMap`.

Reply:
> In fact we don't need to expose hash - we need to suppress it! 

Not really. I'm pretty sure this will not cover all of our needs.

```
type Type1
    Value x y
```

I assume by default the hashcode of `Type1.Value 10 20` will not be equal to `Type1.Value 20 10`, right? Usually the hashcode should be in some way 'positional', as I want such elements to fall to different buckets.

But then, as pointed out by @ekmett IIRC - I may want to provide a hashcode for example of an AVL tree or some other BST such that two trees are equal if they have equal contents (i.e. if their `to_vector` representations are equal). Then I need to be able to 'play' with the ordering so that regardless of the internal shape of the tree the resulting hashcode will be the same (for example, I can traverse the tree in pre-order and aggregate elements' hashcodes).

This kind of usecase is not covered if I were to just rely on `Exclude_From_Equality_And_Hash`.

So I'm still defending the position that we need some way to access the hashcode to be able to write good custom hashcodes for custom notions of equality.

Reply:
## Accessing hashcodes from Java

> Following shall work:
> (...)

You are right, but this is not quite helpful. Sorry - I have initially phrased my previous comment without enough detail.

My point was - if you want to just expose a Map and not a hashcode, how can we use the logic in Java. And yes, your answer shows us how we can use the Map in Java. But that is not really sufficient for us. Big parts of our Table library just need to be access raw hashcodes and taking that away would significantly complicate our code there, if even sensible workarounds were still possible.

I do appreciate the purity of not exposing hashcode on everything, but I think it is the most pragmatic choice and I thought the way to access it through the `comparator` is a compromise we could agree on. Without it, the life of a library developer trying to use hashing will be much harder.

To show a real example from our current codebase, please note the [`MultiValueIndex`](https://github.com/enso-org/enso/blob/e15583fe65b5846f93f8a947a8dd9d6eb538d135/std-bits/table/src/main/java/org/enso/table/data/index/MultiValueIndex.java) which we are using to implement hash-based joins. It relies on the [`UnorderedMultiValueKey`](https://github.com/enso-org/enso/blob/e15583fe65b5846f93f8a947a8dd9d6eb538d135/std-bits/table/src/main/java/org/enso/table/data/index/UnorderedMultiValueKey.java) which is currently using the Java hashcode and will need to be switched to the Enso hashcode once this PR is implemented. Note that the `MultiValueKey` is computed based on hashcodes of multiple values in it, because it is dealing with rows of multiple objects. Having _just_ access to a _Map_ would not be enough to still be able to use this implementation. Of course we could probably emulate it by allocating some scaffolding Enso objects to trigger proper hashing mechanisms, but not only that would be less efficient but also make the code much more complex.

Also this is showing well what we need - we will need a way to access the Enso hashcode of any value in the Java library - and at least for the primitive types (`double`, `long`, `String`, `boolean`) we want this to be as fast as possible. Calling back an Enso function from Java is known to not be so fast, so ideally we should have some Java library accessible from our `std-base`/`std-table` that can provide the hashcode for primitives without that overhead and only delegate to the Enso callback for custom objects.

Reply:
Agree we need some way to get to the hashcode (and equality) for polyglot space. The case Radek mentions is the obvious first place.

I suppose we could have an implementation of MVI built in the Enso verse and pass that in but that feels like it would be slower. Longer term we will probably rebuild Table as a built in but that's down the road.

I also think advanced users will need hashcode access for somethings in polyglot space.